### PR TITLE
Auth/pm-54 add rbac for users

### PR DESCRIPTION
**Pull request by dgoodman-bw**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Add RBAC for the User entity. 

## Code changes

* **UsersController.cs** Replaced the ToUser() function with value assignments dependent on the logged in user's permissions. The permission filter attribute was also applied to the Delete action.
* **Permissions.cs** The names of these permissions were changed to better reflect their purpose.
* **BillingInformationModel.cs** This model and accompanying partial view are used on both the User and Organization pages. An Entity property was added to be able to tell if we're in the User or Organization page so the appropriate permissions can be checked. 
* **UserEditModel.cs** This function is no longer used. Name, Email, and EmailVerified can no longer be updated, and the other value assignments have been moved to the controller where permission checks can be applied. 
* **RolePermissionMapping.cs** The UpgradePremium permission was added to the Billing role. The other edits are just name changes.
* **Organizations/Edit.cshtml** Passes to the partial view that the billing information displayed belongs to an Organization. This information determines what permission will be checked. 
* **_BillingInformation.cshtml** Permission gates have been added to the UI
* **_Layout.cshtml** Remove the "Users" option in the nav bar if the logged in user does not have permission to view the page the option links to.
* **Users/Edit.cshtml** Permission checks have been added to prevent logged in users from seeing or editing information they're not authorized to interact with. 

**Owner Role View**
Owner

**Admin Role View**
Admin

**CS Role View**
CS

**Billing Role View**
Billing

**Sales Role View**
Sales

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Admin/Views/Shared/_BillingInformation.cshtml [hunk 1]
```diff
@@ -1,4 +1,15 @@
-﻿@model BillingInformationModel
+﻿@using Bit.Admin.Enums;
+@inject Bit.Admin.Services.IAccessControlService AccessControlService
+@model BillingInformationModel
+
+@{
+    var canManageTransactions = Model.Entity == "User" ? AccessControlService.UserHasPermission(Permission.User_BillingInformation_CreateEditTransaction) 
+                                                       : AccessControlService.UserHasPermission(Permission.Org_BillingInformation_CreateEditTransaction);
+
+    var canDowloadInvoice = Model.Entity == "User" ? AccessControlService.UserHasPermission(Permission.User_BillingInformation_DownloadInvoice)
```
Comment (JaredSnider-Bitwarden) [inline src/Admin/Views/Shared/_BillingInformation.cshtml:9]: canDowloadInvoice. --> canDow**n**loadInvoice

File: src/Admin/Views/Users/Edit.cshtml [hunk 2]
```diff
@@ -126,39 +152,52 @@
...
                 <label asp-for="GatewayCustomerId"></label>
                 <div class="input-group">
-                    <input type="text" class="form-control" asp-for="GatewayCustomerId">
-                    <div class="input-group-append">
-                        <button class="btn btn-secondary" type="button" id="gateway-customer-link">
-                            <i class="fa fa-external-link"></i>
-                        </button>
-                    </div>
+                    <input type="text" class="form-control" asp-for="GatewayCustomerId" disabled='@(canEditBilling ? null : "disabled")'>
+                    @if (canLaunchInStripe)
```
Comment (JaredSnider-Bitwarden) [inline src/Admin/Views/Users/Edit.cshtml:156]: ❓  the gateway is implemented using a select which implies we can have non-stripe billing gateways, but then we name this permission specifically to be `canLaunchInStripe`?   Do we have non-Stripe billing gateways?
image

**Comment by JaredSnider-Bitwarden** [comment]
Nice work on this!

File: src/Admin/Views/Users/Edit.cshtml (hunk 2 above)
Comment (dgoodman-bw) [reply src/Admin/Views/Users/Edit.cshtml:156]: Great point. We also have Braintree. The requirements specifically called out launching in Stripe, so I'll confirm if the permission check should be applied to just Stripe or to all gateways, and either update the name of the permission or the logic around it depending on the answer.

Comment (JaredSnider-Bitwarden) [reply src/Admin/Views/Users/Edit.cshtml:156]: Awesome. Thank you!

Comment (JaredSnider-Bitwarden) [reply src/Admin/Views/Users/Edit.cshtml:156]: @dgoodman-bw do you want me to go ahead and approve this or wait on this potential refactoring?

Comment (dgoodman-bw) [reply src/Admin/Views/Users/Edit.cshtml]: @JaredSnider-Bitwarden Turns out the permission is intended to include all billing gateways. The change has been implemented and is now ready for re-review. Thank you!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
LGTM 👍 great work on this!

**Pull request merged**